### PR TITLE
Code Refactor: Simpler, Pythonic and more readable

### PR DESCRIPTION
**Pull request by HighnessAtharva**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=G6wTfcq)
Code associated with the following modules refactored to be more readable, easy to understand and follow Pythonic style:
- arithmetic analysis
- audio filters
- backtracking

### Describe your change:

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: arithmetic_analysis/bisection.py [hunk 1]
```diff
@@ -21,13 +21,12 @@ def bisection(function: Callable[[float], float], a: float, b: float) -> float:
     """
     start: float = a
     end: float = b
-    if function(a) == 0:  # one of the a or b is a root for the function
-        return a
+    if function(start) == 0:  # one of the a or b is a root for the function
```
Comment (realstealthninja) [inline arithmetic_analysis/bisection.py:24]:
```suggestion
    if function(start) == 0:  # start or end is a root for the function
```

**Comment by HighnessAtharva** [comment]
I will be refactoring several other files if this PR gets merged to improve code quality and Pythonic style.

File: arithmetic_analysis/jacobi_iteration_method.py [hunk 2]
```diff
@@ -144,20 +144,12 @@ def strictly_diagonally_dominant(table: NDArray[float64]) -> bool:
...
-        for j in range(0, cols - 1):
-            if i == j:
-                continue
-            else:
-                total += table[i][j]
-
+    for i in range(rows):
+        total = sum(table[i][j] for j in range(cols - 1) if i != j)
         if table[i][i] <= total:
             raise ValueError("Coefficient matrix is not strictly diagonally dominant")
```
Comment (tianyizheng02) [inline arithmetic_analysis/jacobi_iteration_method.py:150]:
```suggestion
            return False
```
The function currently returns `True` if `table` is strictly diagonally dominant, but it raises a `ValueError` if `table` isn't. It makes more sense to just have it return `False` instead.

File: backtracking/sudoku.py [hunk 3]
```diff
@@ -51,24 +54,21 @@ def is_safe(grid: Matrix, row: int, column: int, n: int) -> bool:
...
 
-    for i in range(3):
-        for j in range(3):
-            if grid[(row - row % 3) + i][(column - column % 3) + j] == n:
-                return False
-
-    return True
+    return all(
+        grid[(row - row % 3) + i][(column - column % 3) + j] != n
+        for i, j in itertools.product(range(3), range(3))
```
Comment (tianyizheng02) [inline backtracking/sudoku.py:59]:
```suggestion
        for i in range(3) for j in range(3))
```
This should do the same thing without needing an imported function

File: backtracking/sudoku.py [hunk 4]
```diff
@@ -51,24 +54,21 @@ def is_safe(grid: Matrix, row: int, column: int, n: int) -> bool:
...
     This function finds an empty location so that we can assign a number
     for that particular row and column.
     """
-    for i in range(9):
-        for j in range(9):
-            if grid[i][j] == 0:
-                return i, j
-    return None
+    return next(
+        ((i, j) for i, j in itertools.product(range(9), range(9)) if grid[i][j] == 0),
```
Comment (tianyizheng02) [inline backtracking/sudoku.py:69]:
```suggestion
        ((i, j) for i in range(9) for j in range(9) if grid[i][j] == 0),
```
This should do the same thing without needed an imported function

File: backtracking/sudoku.py [hunk 5]
```diff
@@ -9,8 +9,11 @@
 have solved the puzzle. else, we backtrack and place another number
 in that cell and repeat this process.
 """
+
 from __future__ import annotations
 
+import itertools
```
Comment (tianyizheng02) [inline backtracking/sudoku.py:15]:
```suggestion
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]

**Comment by HighnessAtharva** [comment]
Code is messy due to accepted suggestions. Making a fresh PR soon.

**Comment by tianyizheng02** [comment]
```
>>> table = np.array([[4, 1, 1, 2], [1, 5, 2, -6], [1, 2, 3, -4]])
    >>> strictly_diagonally_dominant(table)
    Traceback (most recent call last):
        ...
ValueError: Coefficient matrix is not strictly diagonally dominant
```
@HighnessAtharva I believe this is the one test that's failing because the function now returns `False` instead of raising `ValueError`

**Comment by tianyizheng02** [comment]
> should I reopen the PR and fix it?

Yeah I think the PR should be OK once you fix that one doctest

**Pull request closed without merging**